File: server/src/song/song.service.ts
import {Injectable} from "@nestjs/common";
import {InjectModel} from "@nestjs/mongoose";

import {Model, ObjectId} from "mongoose";

import {Song, SongDocument} from "./schemas/song.schema";
import {Comment, CommentDocument} from "./schemas/comment.schema";
import {CreateSongDto} from "./dto/create-song.dto";
import {CreateCommentDto} from "./dto/create-comment.dto";
import {FileService, FileType} from "../file/file.service";
import {User, UserDocument} from "../users/schemas/user.schema";
import {Album, AlbumDocument} from "../album/schemas/album.schema";


@Injectable()
export class SongService {

    constructor(@InjectModel(Song.name) private songModel: Model<SongDocument>,
                @InjectModel(Comment.name) private commentModel: Model<CommentDocument>,
                @InjectModel(User.name) private userModel: Model<UserDocument>,
                @InjectModel(Album.name) private albumModel: Model<AlbumDocument>,
                private fileService: FileService) {}

    async create(dto: CreateSongDto, user, picture, audio): Promise<Song> {
        const audioPath = this.fileService.createFile(FileType.AUDIO, audio);
        const picturePath = this.fileService.createFile(FileType.IMAGE, picture);
        const updateUser = await this.userModel.findById(user);
        const song = await this.songModel.create({...dto, user: user, listens: 0, audio: audioPath, picture: picturePath});

        updateUser.songs.push(song._id);
        await updateUser.save();

        return song;
    }

    async getAll(count = 10, offset = 0, filter = ''): Promise<Song[]> {
        if (filter === "new"){
            return this.songModel.find().sort([['createdAt', -1]]).skip(Number(offset)).limit(Number(count));
        } else if (filter==="aph"){
            return this.songModel.find().sort([['name', 1]]).skip(Number(offset)).limit(Number(count));
        } else if (filter==="art"){
            return this.songModel.find().sort([['artist', 1]]).skip(Number(offset)).limit(Number(count));
        } else {
            return this.songModel.find().skip(Number(offset)).limit(Number(count));
        }
    }

    async getOne(id: ObjectId): Promise<Song> {
        const song = await this.songModel.findById(id).populate('comments');
        return song;
    }

    async delete(id: ObjectId): Promise<ObjectId> {
        const file = {
            audio: await this.songModel.findById(id).select('audio'),
            picture: await this.songModel.findById(id).select('picture')
        };
        console.log(file);
        const song = await this.songModel.findByIdAndDelete(id);
        const album = await this.albumModel.findById(song.album);
        // delete song from album
        if(album){
            const index = album.songs.indexOf(song._id);
            album.songs.splice(index, 1);
            await album.save();
        }


        await this.fileService.removeFile(file.audio.audio);
        await this.fileService.removeFile(file.picture.picture);
        return song.id;
    }
    async addComment(dto: CreateCommentDto): Promise<Comment> {
        const song = await this.songModel.findById(dto.songId);
        const comment = await this.commentModel.create({...dto})
        song.comments.push(comment._id)
        await song.save();
        return comment;
    }

   async listen(id: ObjectId) {
        const song = await this.songModel.findById(id);
        song.listens++;
        song.save();
    }

   async search(search: string): Promise<Song[]> {
        const [songs] = await Promise.all([
            this.songModel.find({
                $or: [
                    { name: { $regex: new RegExp(search, 'i') } },
                    { artist: { $regex: new RegExp(search, 'i') } },
                ],
            }),
        ]);
        return songs;
    }

    async getCount(){
        const songCount = await this.songModel.countDocuments({}).exec();
        return songCount;
    }

    async getAllUserSongs(id: ObjectId) {
        const songs = await this.songModel.find({
                $or: [
                    {user: id, album: {$exists: false}},
                    {user: id, album: null}
                ]
            }).exec();
        return songs;
    }
}

File: backend/src/main/java/com/muz/service/SongService.java
package com.muz.service;

import com.muz.document.AlbumDocument;
import com.muz.document.CommentDocument;
import com.muz.document.SongDocument;
import com.muz.document.UserDocument;
import com.muz.dto.CreateCommentDto;
import com.muz.dto.CreateSongDto;
import com.muz.exception.ApiException;
import com.muz.repository.AlbumRepository;
import com.muz.repository.CommentRepository;
import com.muz.repository.SongRepository;
import com.muz.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class SongService {

    private final SongRepository songRepository;
    private final CommentRepository commentRepository;
    private final UserRepository userRepository;
    private final AlbumRepository albumRepository;
    private final FileStorageService fileStorage;

    public SongDocument create(CreateSongDto dto, String userId, MultipartFile picture, MultipartFile audio) {
        String picturePath = fileStorage.store("image", picture);
        String audioPath = fileStorage.store("audio", audio);

        SongDocument song = new SongDocument();
        song.setName(dto.getName());
        song.setArtist(dto.getArtist());
        song.setText(dto.getText());
        song.setPicture(picturePath);
        song.setAudio(audioPath);
        song.setUser(userId);
        SongDocument saved = songRepository.save(song);

        UserDocument user = userRepository.findById(userId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "User not found"));
        user.getSongs().add(saved.getId());
        userRepository.save(user);

        return saved;
    }

    public List<SongDocument> getAll(int count, int offset, String filter) {
        PageRequest page = PageRequest.of(offset / count, count);
        return switch (filter) {
            case "new" -> songRepository.findAllByOrderByCreatedAtDesc(page);
            case "aph" -> songRepository.findAllByOrderByNameAsc(page);
            case "art" -> songRepository.findAllByOrderByArtistAsc(page);
            default -> songRepository.findAll(page).getContent();
        };
    }

    public Map<String, Object> getOne(String id) {
        SongDocument song = songRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Song not found"));
        List<CommentDocument> comments = commentRepository.findAllByIdIn(song.getComments());
        return Map.of("song", song, "comments", comments);
    }

    public String delete(String id) {
        SongDocument song = songRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Song not found"));

        if (song.getAlbum() != null) {
            albumRepository.findById(song.getAlbum()).ifPresent(album -> {
                album.getSongs().remove(song.getId());
                albumRepository.save(album);
            });
        }

        songRepository.deleteById(id);
        fileStorage.delete(song.getAudio());
        fileStorage.delete(song.getPicture());
        return id;
    }

    public CommentDocument addComment(CreateCommentDto dto) {
        SongDocument song = songRepository.findById(dto.getSongId())
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Song not found"));

        CommentDocument comment = new CommentDocument();
        comment.setUsername(dto.getUsername());
        comment.setText(dto.getText());
        comment.setTrack(dto.getSongId());
        CommentDocument saved = commentRepository.save(comment);

        song.getComments().add(saved.getId());
        songRepository.save(song);
        return saved;
    }

    public void listen(String id) {
        SongDocument song = songRepository.findById(id)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Song not found"));
        song.setListens(song.getListens() + 1);
        songRepository.save(song);
    }

    public List<SongDocument> search(String term) {
        return songRepository.searchByNameOrArtist(term);
    }

    public long getCount() {
        return songRepository.count();
    }

    public List<SongDocument> getAllUserSongs(String userId) {
        return songRepository.findUserSongsWithoutAlbum(userId);
    }
}
